### ISSUE TITLE
I18n support

### ISSUE THREAD

**Issue by niksy** [date-fns/date-fns#157]
Are there any plans to have support for custom locale for `format` module? It would be nice to have this for other languages.

**Comment by kossnocorp** [comment]
@niksy currently we are focused on v1.0.0 (couple weeks at max), but the next focus is locales. If you want to help us with Croatian, we can connect, and I'll explain how you can do it.

**Comment by mikkelwf** [comment]
I would like to help with danish translation, when you get to it.. :)

**Comment by kossnocorp** [comment]
@mikkelwf thank you, it would be really helpful!

@niksy I'm going to reuse thus issue, I hope you are not against 😄

**Comment by just-boris** [comment]
Hello! I also need this feature, so I can contribute it.

I found two locale dependent modules so far.
* [distance_in_words](https://github.com/date-fns/date-fns/blob/master/src/distance_in_words/index.js#L176)
* [format](https://github.com/date-fns/date-fns/blob/master/src/format/index.js#L345)

So, I propose the following solution:
1. Move all locale data in separate module `date-fns/locale`. All functions will import values from there.
2. Expose a method to replace locale

```js
import {replaceLocale} from 'date-fns/locale';

replaceLocale({
  months: ['Januar', 'Februar', 'März', 'April', 'Mai', 'Juni', 'Juli', 'August', 'September', 'Oktober', 'November', 'Dezember']
});
```
That method will reassign internal locale keys, which is used by functions.

What do you think about that API?

**Comment by kossnocorp** [comment]
@just-boris I prefer to keep functions pure (as date functions could be pure), so instead of monkey patching internals by calling `useLocale` we going to allow pass a locale object as an argument.

You can take a look at my [initial idea of API in this comment](https://github.com/date-fns/date-fns/issues/203#issuecomment-220386905). 

[Initial implementation is done](https://github.com/date-fns/date-fns/compare/i18n) and ready for translation to different languages. Currently, I stuck (burnt out) on Russian translation. @LeshaKoss wanted to continue my work on that. 

If you want to join the forces, you can take a look at [the English locale](https://github.com/date-fns/date-fns/tree/8eab7d9306e5c549d8d3c7f59255949b68dccf31/src/locale/en) or wait for the translation manual.

Thank you for your contribution!

**Comment by MarcCloud** [comment]
Hi, I'd like to contribute with the translation to Spanish.

**Comment by nktssh** [comment]
Hi, I'd like to contribute with the translations too!
I think, translations (at least for days/months etc. naming) can be used from [moment.js code](https://github.com/moment/moment/blob/develop/src/locale/)

@kossnocorp @LeshaKoss
May I help you somehow with Russian translations?

**Comment by just-boris** [comment]
So far I have found a way how can I use native browser features to localize date.
When I need to print month in user's locale, I do the following

```
new Date().toLocaleString(navigator.language, {month: 'long'}) // prints Oktober
```
Same for weekdays
```
new Date().toLocaleString(navigator.language, {weekday: 'long'}) // prints Mittwoch
```

So we don't need to store translations somewhere in library or external plugins. It can be just extracted from the browser.

**Comment by kossnocorp** [comment]
Guys, I18n is now supported. It was released as `1.5.2`. See docs: https://date-fns.org/docs/I18n

**Comment by kossnocorp** [comment]
If anyone here wants to contribute, you're more than welcome: https://date-fns.org/docs/I18n#adding-new-langugage. A better guide is coming soon.
